src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -19,208 +19,267 @@ namespace Castle.MicroKernel
 		}
 
-		bool IDictionary.IsReadOnly
+		void ICollection.CopyTo(Array array, int index)
 		{
-			get { return arguments.IsReadOnly; }
+			dictionary.CopyTo(array, index);
 		}
 
-		public void Add(object key, object value)
+		public bool Contains(object key)
 		{
-			arguments.Add(key, value);
+			return dictionary.Contains(key);
 		}
 
-		public void Clear()
+		IEnumerator IEnumerable.GetEnumerator()
 		{
-			arguments.Clear();
+			return GetEnumerator();
 		}
 
-		public bool Contains(object key)
+		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.Contains(key);
+			return dictionary.GetEnumerator();
 		}
 
-		public IDictionaryEnumerator GetEnumerator()
+		/// <summary>
+		/// Inserts a new named argument with a given dependency. If an argument for this named key already exists, it will be overwritten.
+		/// </summary>
+		/// <param named="namedOrTyped">Named parameter</param>
+		/// <param named="value">Instance dependency</param>
+		public virtual Arguments Insert(string named, object value)
 		{
-			return arguments.GetEnumerator();
+			this[named] = value;
+			return this;
 		}
 
-		public void Remove(object key)
+		/// <summary>
+		/// Inserts dictionary with named/typed arguments. If an argument for this named/typed key already exists, it will be overwritten.
+		/// </summary>
+		/// <param named="values"></param>
+		/// <returns><see cref="Arguments"/></returns>
+		public virtual Arguments Insert(IDictionary values)
 		{
-			arguments.Remove(key);
+			foreach (DictionaryEntry entry in values)
+			{
+				if (entry.Key is string || entry.Key is Type)
+				{
+					dictionary[entry.Key] = entry.Value;
+				}
+				else
+				{
+					throw new ArgumentException($"The argument namedOrTyped '{entry.Key}' should be of type System.String or System.Type");
+				}
+			}
+			return this;
 		}
 
-		public Arguments Clone()
+		/// <summary>
+		/// Method that can be used to named arguments into <see cref="Arguments"/> from a read-only dictionary.
+		/// </summary>
+		/// <param named="values"><see cref="IReadOnlyDictionary{S, O}"/> where the dictionary is shallow copied up front into <see cref="Arguments"/> and then used as dependencies</param>
+		/// <returns><see cref="Arguments"/></returns>
+		public virtual Arguments InsertNamed(IReadOnlyDictionary<string, object> values)
 		{
-			var dictionary = arguments as Dictionary<object, object>;
-			if (dictionary != null)
+			isReadOnly = false;
+			foreach (var entry in values)
 			{
-				var comparerExtended = dictionary.Comparer as ArgumentsComparerExtended;
-				if (comparerExtended != null)
+				if (entry.Key is string)
 				{
-					return new Arguments(arguments, comparerExtended.CustomComparers);
+					dictionary[entry.Key] = entry.Value;
+				}
+				else
+				{
+					throw new ArgumentException($"The argument namedOrTyped '{entry.Key}' should be of type System.String");
 				}
 			}
 
-			return new Arguments(arguments);
+			isReadOnly = true;
```

Comment:
I thought we were on the same page about `IReadOnlyDictionary` being handled just like any other dictionary, the `Arguments` instance shouldn't be made read-only just because you insert some items from a collection implementing a read-only view? This behaviour is non-obvious and doesn't align with the behaviour of the other Insert overloads.

It now looks like the only use for the `isReadOnly` flag is the `Empty` instance, but that might change when we look at (post v5) the next part when the `Arguments` instance is passed to Windsor for resolution, I suspect Windsor will probably freeze/make-read-only the instance before it starts using it.

Reply:
> I thought we were on the same page about IReadOnlyDictionary being handled just like any other dictionary, the Arguments instance shouldn't be made read-only just because you insert some items from a collection implementing a read-only view? This behaviour is non-obvious and doesn't align with the behaviour of the other Insert overloads.

Not too sure about that. I would like to see that read only contract honoured throughout the Windsor code base(incl. Facilities). If something is mutating it I would rather have tests fail on PR's. We simply cannot accept mutative behaviour downstream for anyone that uses this API. Ideally for performance, we would later on eliminate the "copy up front" approach.

Reply:
> I would like to see that read only contract honoured

I think this is our biggest disagreement in this pull request. `IReadOnlyDictionary` doesn't define an immutable contract, it defines a readable view of a dictionary, nothing stops the dictionary changing from when you were given a reference to the view. Mentally I think of it as `IReadableDictionary`.

I don't see why the following code shouldn't be allowed:

```csharp
new Arguments("first", "value")
    .Insert("key", "value")
    .Insert(new Dictionary<string, object>())
    .Insert("third", "value"); // <- throws NotSupportedException: Collection is read-only
```